### ISSUE TITLE
Issue with the 'na.rm' argument in dplyr::n_distinct

### ISSUE THREAD

**Issue by xingzhong-jin** [sparklyr/sparklyr#3177]
---

The calculation of `dplyr::n_distinct` applied to a spark data frame is different from the result when it is applied locally. The `na.rm = TRUE` argument is ignored when applied to a spark data frame. 

```r
library(sparklyr)
library(dplyr)

spark_version <- "3.0.0"

sc <- spark_connect(master = "local", version = spark_version)

# example data
x <- rep(1:3, 2)
y <- c(rep("a", 3), rep(NA, 3))  

df_local <- data.frame(x, y) 

df_spark <- copy_to(sc, df_local, name = "df_spark")

# example n_distinct function
n_distinct_y_by_x <- function(df) {
  df %>% 
    group_by(x) %>% 
    summarise(n_row = n(), 
              n_unique = n_distinct(y, na.rm = FALSE), 
              n_na = n_distinct(y, na.rm = FALSE) - n_distinct(y, na.rm = TRUE)) %>% 
    ungroup() %>% 
    arrange(x)
  
}
```

```r
# local dataframe 

df_local %>% 
  n_distinct_y_by_x()
```
> A tibble: 3 × 4
      x n_row n_unique  n_na
  <int> <int>    <int> <int>
1     1     2        2     1
2     2     2        2     1
3     3     2        2     1

```r
# spark dataframe 

df_spark %>% 
  n_distinct_y_by_x()

```
>Source:     spark<?> [?? x 4]
 Ordered by: x
      x n_row n_unique  n_na
  <int> <dbl>    <dbl> <dbl>
1     1     2        2     0
2     2     2        2     0
3     3     2        2     0

```r
sessionInfo()
```

>R version 3.6.0 (2019-04-26)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.16
Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     
other attached packages:
[1] dplyr_1.0.7    sparklyr_1.7.1
loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       pillar_1.6.2     compiler_3.6.0   dbplyr_2.1.1     r2d3_0.2.3      
 [6] base64enc_0.1-3  tools_3.6.0      digest_0.6.19    jsonlite_1.6     lifecycle_1.0.0 
[11] tibble_3.1.3     pkgconfig_2.0.2  rlang_0.4.11     cli_3.0.1        DBI_1.0.0       
[16] rstudioapi_0.13  curl_3.3         yaml_2.2.0       parallel_3.6.0   withr_2.4.0     
[21] httr_1.4.1       generics_0.1.0   vctrs_0.3.8      htmlwidgets_1.3  askpass_1.1     
[26] rappdirs_0.3.1   rprojroot_2.0.2  tidyselect_1.1.0 glue_1.4.2       forge_0.2.0     
[31] R6_2.4.0         fansi_0.4.0      blob_1.2.1       purrr_0.3.4      tidyr_1.1.3     
[36] magrittr_1.5     ellipsis_0.3.2   htmltools_0.3.6  assertthat_0.2.1 config_0.3      
[41] utf8_1.1.4       openssl_1.4.3    crayon_1.3.4

**Comment by yitao-li** [comment]
`n_distinct(<expr>)` is translated simply into "count(distinct(<SQL expr>))` at the moment, and `na.rm` is not supported yet.

**Comment by yitao-li** [comment]
Feature request: https://github.com/sparklyr/sparklyr/issues/3178

**Comment by yitao-li** [comment]
@xingzhong-jin While the `na.rm` option is not supported yet by `sparklyr`at the moment, I thought of a temp workaround that would have the same effect as `na.rm = TRUE` in base R dataframes.

The workaround: replace `n_distinct(y, na.rm = FALSE)` in your query with `n_distinct(array(if(isnan(y), NULL, y)))`.
Why it will work:
- Once `if(isnan(y), NULL, y)` is wrapped inside a Spark array, the result is an array, and therefore will definitely be not `NULL` (well, obviously :slightly_smiling_face: )
- 2 or more instances of arrays containing a single null will reduce to a single unique value after "distinct"
- Finally, in Spark SQL it is possible for a numeric column to have `NULL` *or* `NaN` (and the latter is considered not `NULL`). But to emulate the `na.rm` behavior in base R, `NaN` value should be treated the same as `NULL` value for numeric expressions, hence `n_distinct()` is applied to `if(isnan(y), NULL, y)` rather than just applied to `y`.

**Comment by yitao-li** [comment]
Update: https://github.com/sparklyr/sparklyr/issues/3178 is merged and the change was released on CRAN as part of `sparklyr` 1.7.2.
